File: src/obj-c app/ViewController.m
//
//  ViewController.m
//  sockets
//
//  Created by Syritx on 2021-01-20.
//

#import "ViewController.h"
#import <CFNetwork/CFNetwork.h>
#import <Foundation/Foundation.h>
#import <stdio.h>
#import <string.h>
#import <stdlib.h>
#import <unistd.h>
#import <pthread.h>
#import <sys/socket.h>
#import <sys/un.h>
#import <sys/stat.h>
#import <sys/types.h>
#import <netinet/in.h>
#import <UIKit/UIKit.h>
#include <arpa/inet.h>

@interface ViewController ()

@end

@implementation ViewController

- (IBAction) sendMessageToServer : (id)sender {
    NSLog(@"sending");
    NSData *data = [[NSData alloc] initWithData:[@"[get_data]" dataUsingEncoding:NSASCIIStringEncoding]];
    [outputStream write:[data bytes] maxLength:[data length]];
    
    
    uint8_t buffer[1024];
    int len;
    while ([inputStream hasBytesAvailable]) {
        len = [inputStream read:buffer maxLength:sizeof(buffer)];
        if (len > 0) {
            NSLog(@"got data");
            NSString *output = [[NSString alloc] initWithBytes:buffer length:len encoding:NSASCIIStringEncoding];
            if (output != nil) {
                NSLog(@"%@", output);
                NSArray* COMMANDS = [output componentsSeparatedByString: @"[command_buffer]"];
                for (int i = 0; i < [COMMANDS count]; i++) {
                    NSLog(@"%@ index: %d", [COMMANDS objectAtIndex: i], i);
                    if ([[COMMANDS objectAtIndex: i] hasPrefix: @"[temperature]:"]) {
                        [temperatureLabel setText: [[COMMANDS objectAtIndex: i] uppercaseString]];
                    }
                    if ([[COMMANDS objectAtIndex: i] hasPrefix: @"[humidity]:"]) {
                        [humidityLabel setText: [[COMMANDS objectAtIndex: i] uppercaseString]];
                    }
                }
            }
        }
    }
}

- (IBAction) connectToServer : (id)sender {
    
    // CHANGE 'HOST' VARIABLE DECLARED AT LINE 96
    CFStreamCreatePairWithSocketToHost(kCFAllocatorDefault, (__bridge CFStringRef)HOST, 6060, &readStream, &writeStream);
    
    inputStream = (__bridge NSInputStream *)readStream;
    outputStream = (__bridge NSOutputStream *)writeStream;
    [inputStream setDelegate:self];
    [outputStream setDelegate:self];
    [inputStream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
    [outputStream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
    [inputStream open];
    [outputStream open];
    
    NSData *data = [[NSData alloc] initWithData:[@"IPhone app connected" dataUsingEncoding:NSASCIIStringEncoding]];
    [outputStream write:[data bytes] maxLength:[data length]];
    
    
    NSLog(@"here");
}

- (IBAction) disconnectFromServer : (id)sender {
    NSLog(@"sending");
    NSData *data = [[NSData alloc] initWithData:[@"disconnect_ed" dataUsingEncoding:NSASCIIStringEncoding]];
    [outputStream write:[data bytes] maxLength:[data length]];
    [outputStream close];
}

void getSocketCallBack(CFSocketRef cfSocketRef, CFSocketCallBackType cbType, CFDataRef address, const void *data, void *info) {

}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // IMPORTANT //
    HOST = @"ip-here"; // i.e. X.X.X.X
    
    UIView *mainview;
    [mainview setBackgroundColor: [UIColor blackColor]];
    
    float width = UIScreen.mainScreen.bounds.size.width-40;
    float yOffset = 20;
    float inlineOffset = 10;
    
    UIButton *messageButton = [[UIButton alloc] initWithFrame:CGRectMake(20, 20+yOffset+inlineOffset, width, 50)];
    [messageButton setBackgroundColor:[UIColor linkColor]];
    [messageButton addTarget:self action:@selector(sendMessageToServer:) forControlEvents:UIControlEventTouchUpInside];
    [messageButton setTitle:@"Request Data" forState:UIControlStateNormal];
    
    UIButton *connect = [[UIButton alloc] initWithFrame: CGRectMake(20, 70+yOffset+inlineOffset*2, width, 50)];
    [connect setBackgroundColor:[UIColor linkColor]];
    [connect addTarget:self action:@selector(connectToServer:) forControlEvents:UIControlEventTouchUpInside];
    [connect setTitle:@"Connect" forState:UIControlStateNormal];
    
    UIButton *disconnect = [[UIButton alloc] initWithFrame: CGRectMake(20, 120+yOffset+inlineOffset*3, width, 50)];
    [disconnect setBackgroundColor:[UIColor linkColor]];
    [disconnect addTarget:self action:@selector(disconnectFromServer:) forControlEvents:UIControlEventTouchUpInside];
    [disconnect setTitle:@"Disconnect" forState:UIControlStateNormal];
    
    temperatureLabel = [[UILabel alloc] initWithFrame: CGRectMake(20, 170+yOffset+inlineOffset*4, width, 50)];
    [temperatureLabel setText: @"[TEMPERATURE]: "];
    [temperatureLabel setTextColor: [UIColor whiteColor]];
    [temperatureLabel setBackgroundColor: [UIColor blackColor]];
    
    humidityLabel = [[UILabel alloc] initWithFrame: CGRectMake(20, 200+yOffset+inlineOffset*5, width, 50)];
    [humidityLabel setText: @"[HUMIDITY]: "];
    [humidityLabel setTextColor: [UIColor whiteColor]];
    [humidityLabel setBackgroundColor: [UIColor blackColor]];
    
    [self.view setBackgroundColor: [UIColor blackColor]];
    [self.view addSubview: mainview];
    [self.view addSubview: messageButton];
    [self.view addSubview: connect];
    [self.view addSubview: disconnect];
    [self.view addSubview: temperatureLabel];
    [self.view addSubview: humidityLabel];
}

@end

File: src/swift app/ViewController.swift
//
//  ViewController.swift
//  swift-sockets
//
//  Created by Syritx on 2021-01-21.
//

import UIKit
import Foundation

class ViewController: UIViewController {
    
    var readStream : Unmanaged<CFReadStream>?
    var writeStream : Unmanaged<CFWriteStream>?
    
    var inputStream : InputStream!
    var outputStream : OutputStream!
    
    var temperatureLabel : UILabel!
    var humidityLabel : UILabel!
    
    var HOST : NSString?
    
    override func viewDidLoad() {
        super.viewDidLoad()
            
        HOST = "192.168.0.165"
        let uiView = UIView()
        uiView.backgroundColor = .black
        
        let width = UIScreen.main.bounds.width-40
        let yOffset = 20
        let inlineOffset = 10
        
        let messageButton = UIButton()
        messageButton.frame = CGRect(x: 20, y: 20+yOffset+inlineOffset, width: Int(width), height: 50)
        messageButton.setTitle("Request Data", for: .normal)
        messageButton.backgroundColor = .link
        messageButton.setTitleColor(.white, for: .normal)
        messageButton.addTarget(self, action: #selector(requestDataFromServer(_:)), for: .touchUpInside)
        
        let disconnect = UIButton()
        disconnect.frame = CGRect(x: 20, y: 120+yOffset+inlineOffset*3, width: Int(width), height: 50)
        disconnect.setTitle("Disconnect", for: .normal)
        disconnect.backgroundColor = .link
        disconnect.setTitleColor(.white, for: .normal)
        disconnect.addTarget(self, action: #selector(disconnectFromServer(_:)), for: .touchUpInside)
        
        let connect = UIButton()
        connect.frame = CGRect(x: 20, y: 70+yOffset+inlineOffset*2, width: Int(width), height: 50)
        connect.setTitle("Connect", for: .normal)
        connect.backgroundColor = .link
        connect.setTitleColor(.white, for: .normal)
        connect.addTarget(self, action: #selector(connectToServer(_:)), for: .touchUpInside)
        
        temperatureLabel = UILabel()
        temperatureLabel.frame = CGRect(x: 20, y: 170+yOffset+inlineOffset*4, width: Int(width), height: 50)
        temperatureLabel.text = "[TEMPERATURE]: 0.0"
        temperatureLabel.textColor = .white
        temperatureLabel.backgroundColor = .black
        
        humidityLabel = UILabel()
        humidityLabel.frame = CGRect(x: 20, y: 200+yOffset+inlineOffset*5, width: Int(width), height: 50)
        humidityLabel.text = "[HUMIDITY]: 0.0"
        humidityLabel.textColor = .white
        humidityLabel.backgroundColor = .black
        
        self.view = uiView
        view.addSubview(messageButton)
        view.addSubview(disconnect)
        view.addSubview(connect)
        view.addSubview(temperatureLabel)
        view.addSubview(humidityLabel)
    }
    
    @objc func requestDataFromServer(_ sender: UIButton) {
        
        let dat = "[get_data]".data(using: .utf8)!
        
        dat.withUnsafeBytes {
            guard let pointer = $0.baseAddress?.assumingMemoryBound(to: UInt8.self)
            else {
                return
            }
            outputStream!.write(pointer, maxLength: dat.count)
        }
    }
    
    @objc func connectToServer(_ sender: UIButton) {
        CFStreamCreatePairWithSocketToHost(kCFAllocatorDefault, HOST, 6060, &readStream, &writeStream)
        inputStream = readStream!.takeRetainedValue()
        outputStream = writeStream!.takeRetainedValue()
        
        inputStream.delegate = self
        inputStream.schedule(in: .current, forMode: .common)
        outputStream.schedule(in: .current, forMode: .common)
        inputStream.open()
        outputStream.open()
    }
    
    @objc func disconnectFromServer(_ sender: UIButton) {
        let dat = "disconnect_ed".data(using: .utf8)!
        
        dat.withUnsafeBytes {
            guard let pointer = $0.baseAddress?.assumingMemoryBound(to: UInt8.self)
            else {
                return
            }
            outputStream!.write(pointer, maxLength: dat.count)
        }
        outputStream!.close()
    }
}

extension ViewController : StreamDelegate {
    
    
    func stream(_ aStream: Stream, handle eventCode: Stream.Event) {
        
        switch eventCode {
            
        case .hasBytesAvailable:
            print("message")
            getReceivedMessage(stream: aStream as! InputStream)
            
        default:
            print("others")
        }
    }
    
    func getReceivedMessage(stream: InputStream) {
        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: 1024)
        var totalLength = 0
        
        while stream.hasBytesAvailable {
            let len = stream.read(buffer, maxLength: 1024)
            if len < 0, let error = stream.streamError {
                print(error)
                break
            }
            print("lol a test lol")
            totalLength = len
            break
        }
        
        print("length is bigger than 0");
        guard let output = String(bytesNoCopy: buffer,
                            length: totalLength,
                            encoding: .utf8,
                            freeWhenDone: true)
        else {
            return
        }
        print(totalLength)
        
        if output.starts(with: "[command_buffer]") {
            print(output)
            let commands = output.components(separatedBy: "[command_buffer]")
            for word in commands {
                print(word)
                if word.starts(with: "[temperature]:") {
                    temperatureLabel.text = word.uppercased()
                }
                
                if word.starts(with: "[humidity]:") {
                    humidityLabel.text = word.uppercased()
                }
            }
        }
    }
}
